File: go-client-sdk/pkg/models/shared/invoice.go
package shared

import (
	"time"
)

// InvoiceAutoCollection
// Information about payment auto-collection for this invoice.
type InvoiceAutoCollection struct {
	NextAttemptAt         *time.Time `json:"next_attempt_at,omitempty"`
	PreviouslyAttemptedAt *time.Time `json:"previously_attempted_at,omitempty"`
}

// InvoiceCustomer
// The customer receiving this invoice.
type InvoiceCustomer struct {
	ExternalCustomerID string `json:"external_customer_id"`
	ID                 string `json:"id"`
}

// InvoiceLineItemsSubLineItemsGrouping
// For configured prices that are split by a grouping key, this will be populated with the key and a value. The `amount` will be the values for this particular grouping.
type InvoiceLineItemsSubLineItemsGrouping struct {
	Key   *string `json:"key,omitempty"`
	Value *string `json:"value,omitempty"`
}

// InvoiceLineItemsSubLineItemsMatrixConfig
// Only available if `type` is `matrix`. Contains the values of the matrix that this `sub_line_item` represents.
type InvoiceLineItemsSubLineItemsMatrixConfig struct {
	DimensionValues []string `json:"dimension_values"`
}

// InvoiceLineItemsSubLineItemsTierConfig
// Only available if `type` is `tier`. Contains the range of units in this tier and the unit amount.
type InvoiceLineItemsSubLineItemsTierConfig struct {
	FirstUnit  float64 `json:"first_unit"`
	LastUnit   float64 `json:"last_unit"`
	UnitAmount string  `json:"unit_amount"`
}

type InvoiceLineItemsSubLineItemsTypeEnum string

const (
	InvoiceLineItemsSubLineItemsTypeEnumMatrix InvoiceLineItemsSubLineItemsTypeEnum = "matrix"
	InvoiceLineItemsSubLineItemsTypeEnumTier   InvoiceLineItemsSubLineItemsTypeEnum = "tier"
)

type InvoiceLineItemsSubLineItems struct {
	Amount       string                                    `json:"amount"`
	Grouping     InvoiceLineItemsSubLineItemsGrouping      `json:"grouping"`
	MatrixConfig *InvoiceLineItemsSubLineItemsMatrixConfig `json:"matrix_config,omitempty"`
	Name         string                                    `json:"name"`
	Quantity     float64                                   `json:"quantity"`
	TierConfig   *InvoiceLineItemsSubLineItemsTierConfig   `json:"tier_config,omitempty"`
	Type         InvoiceLineItemsSubLineItemsTypeEnum      `json:"type"`
}

type InvoiceLineItemsTaxAmounts struct {
	Amount             string `json:"amount"`
	TaxRateDescription string `json:"tax_rate_description"`
	TaxRatePercentage  string `json:"tax_rate_percentage"`
}

type InvoiceLineItems struct {
	Amount       string                         `json:"amount"`
	Discount     map[string]interface{}         `json:"discount"`
	EndDate      time.Time                      `json:"end_date"`
	Grouping     string                         `json:"grouping"`
	ID           string                         `json:"id"`
	Minimum      map[string]interface{}         `json:"minimum"`
	Name         string                         `json:"name"`
	Price        Price                          `json:"price"`
	Quantity     float64                        `json:"quantity"`
	StartDate    time.Time                      `json:"start_date"`
	SubLineItems []InvoiceLineItemsSubLineItems `json:"sub_line_items"`
	Subtotal     string                         `json:"subtotal"`
	TaxAmounts   []InvoiceLineItemsTaxAmounts   `json:"tax_amounts"`
}

type InvoiceStatusEnum string

const (
	InvoiceStatusEnumIssued InvoiceStatusEnum = "issued"
	InvoiceStatusEnumPaid   InvoiceStatusEnum = "paid"
	InvoiceStatusEnumSynced InvoiceStatusEnum = "synced"
	InvoiceStatusEnumVoid   InvoiceStatusEnum = "void"
	InvoiceStatusEnumDraft  InvoiceStatusEnum = "draft"
)

// InvoiceSubscription
// The associated subscription for this invoice.
type InvoiceSubscription struct {
	ID string `json:"id"`
}

// Invoice
// An [`Invoice`](../reference/Orb-API.json/components/schemas/Invoice) is a fundamental billing entity, representing the request for payment for a single subscription. This includes a set of line items, which correspond to prices in the subscription's plan and can represent fixed recurring fees or usage-based fees. They are generated at the end of a billing period, or as the result of an action, such as a cancellation.
type Invoice struct {
	AmountDue                   string                       `json:"amount_due"`
	AutoCollection              *InvoiceAutoCollection       `json:"auto_collection,omitempty"`
	CreatedAt                   time.Time                    `json:"created_at"`
	Currency                    string                       `json:"currency"`
	Customer                    InvoiceCustomer              `json:"customer"`
	CustomerBalanceTransactions []CustomerBalanceTransaction `json:"customer_balance_transactions"`
	CustomerTaxID               *CustomerTaxID               `json:"customer_tax_id,omitempty"`
	Discount                    map[string]interface{}       `json:"discount"`
	DueDate                     time.Time                    `json:"due_date"`
	HostedInvoiceURL            *string                      `json:"hosted_invoice_url,omitempty"`
	ID                          string                       `json:"id"`
	InvoiceDate                 time.Time                    `json:"invoice_date"`
	InvoicePdf                  string                       `json:"invoice_pdf"`
	IssuedAt                    *time.Time                   `json:"issued_at,omitempty"`
	LineItems                   []InvoiceLineItems           `json:"line_items"`
	Memo                        *string                      `json:"memo,omitempty"`
	Minimum                     map[string]interface{}       `json:"minimum"`
	PaidAt                      *time.Time                   `json:"paid_at,omitempty"`
	ScheduledIssueAt            *time.Time                   `json:"scheduled_issue_at,omitempty"`
	Status                      InvoiceStatusEnum            `json:"status"`
	Subscription                InvoiceSubscription          `json:"subscription"`
	Subtotal                    string                       `json:"subtotal"`
	Total                       string                       `json:"total"`
	VoidedAt                    *time.Time                   `json:"voided_at,omitempty"`
}

File: python-client-sdk/src/orbapi/models/shared/invoice.py
from __future__ import annotations
import dataclasses
import dateutil.parser
from ..shared import customer_balance_transaction as shared_customer_balance_transaction
from ..shared import customer_tax_id as shared_customer_tax_id
from ..shared import price as shared_price
from dataclasses_json import Undefined, dataclass_json
from datetime import datetime
from enum import Enum
from marshmallow import fields
from orbapi import utils
from typing import Any, Optional


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceAutoCollection:
    r"""InvoiceAutoCollection
    Information about payment auto-collection for this invoice.
    """
    
    next_attempt_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('next_attempt_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    previously_attempted_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('previously_attempted_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceCustomer:
    r"""InvoiceCustomer
    The customer receiving this invoice.
    """
    
    external_customer_id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('external_customer_id') }})
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItemsSubLineItemsGrouping:
    r"""InvoiceLineItemsSubLineItemsGrouping
    For configured prices that are split by a grouping key, this will be populated with the key and a value. The `amount` will be the values for this particular grouping.
    """
    
    key: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('key'), 'exclude': lambda f: f is None }})
    value: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('value'), 'exclude': lambda f: f is None }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItemsSubLineItemsMatrixConfig:
    r"""InvoiceLineItemsSubLineItemsMatrixConfig
    Only available if `type` is `matrix`. Contains the values of the matrix that this `sub_line_item` represents.
    """
    
    dimension_values: list[str] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('dimension_values') }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItemsSubLineItemsTierConfig:
    r"""InvoiceLineItemsSubLineItemsTierConfig
    Only available if `type` is `tier`. Contains the range of units in this tier and the unit amount.
    """
    
    first_unit: float = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('first_unit') }})
    last_unit: float = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('last_unit') }})
    unit_amount: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('unit_amount') }})
    
class InvoiceLineItemsSubLineItemsTypeEnum(str, Enum):
    MATRIX = "matrix"
    TIER = "tier"


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItemsSubLineItems:
    amount: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('amount') }})
    grouping: InvoiceLineItemsSubLineItemsGrouping = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('grouping') }})
    name: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('name') }})
    quantity: float = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('quantity') }})
    type: InvoiceLineItemsSubLineItemsTypeEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('type') }})
    matrix_config: Optional[InvoiceLineItemsSubLineItemsMatrixConfig] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('matrix_config'), 'exclude': lambda f: f is None }})
    tier_config: Optional[InvoiceLineItemsSubLineItemsTierConfig] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('tier_config'), 'exclude': lambda f: f is None }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItemsTaxAmounts:
    amount: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('amount') }})
    tax_rate_description: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('tax_rate_description') }})
    tax_rate_percentage: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('tax_rate_percentage') }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceLineItems:
    amount: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('amount') }})
    discount: dict[str, Any] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('discount') }})
    end_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('end_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    grouping: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('grouping') }})
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    minimum: dict[str, Any] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('minimum') }})
    name: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('name') }})
    price: shared_price.Price = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('price') }})
    quantity: float = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('quantity') }})
    start_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('start_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    sub_line_items: list[InvoiceLineItemsSubLineItems] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('sub_line_items') }})
    subtotal: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('subtotal') }})
    tax_amounts: list[InvoiceLineItemsTaxAmounts] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('tax_amounts') }})
    
class InvoiceStatusEnum(str, Enum):
    ISSUED = "issued"
    PAID = "paid"
    SYNCED = "synced"
    VOID = "void"
    DRAFT = "draft"


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class InvoiceSubscription:
    r"""InvoiceSubscription
    The associated subscription for this invoice.
    """
    
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class Invoice:
    r"""Invoice
    An [`Invoice`](../reference/Orb-API.json/components/schemas/Invoice) is a fundamental billing entity, representing the request for payment for a single subscription. This includes a set of line items, which correspond to prices in the subscription's plan and can represent fixed recurring fees or usage-based fees. They are generated at the end of a billing period, or as the result of an action, such as a cancellation.
    """
    
    amount_due: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('amount_due') }})
    created_at: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('created_at'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    currency: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('currency') }})
    customer: InvoiceCustomer = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('customer') }})
    customer_balance_transactions: list[shared_customer_balance_transaction.CustomerBalanceTransaction] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('customer_balance_transactions') }})
    discount: dict[str, Any] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('discount') }})
    due_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('due_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    invoice_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('invoice_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    invoice_pdf: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('invoice_pdf') }})
    line_items: list[InvoiceLineItems] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('line_items') }})
    minimum: dict[str, Any] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('minimum') }})
    status: InvoiceStatusEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('status') }})
    subscription: InvoiceSubscription = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('subscription') }})
    subtotal: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('subtotal') }})
    total: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('total') }})
    auto_collection: Optional[InvoiceAutoCollection] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('auto_collection'), 'exclude': lambda f: f is None }})
    customer_tax_id: Optional[shared_customer_tax_id.CustomerTaxID] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('customer_tax_id'), 'exclude': lambda f: f is None }})
    hosted_invoice_url: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('hosted_invoice_url'), 'exclude': lambda f: f is None }})
    issued_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('issued_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    memo: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('memo'), 'exclude': lambda f: f is None }})
    paid_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('paid_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    scheduled_issue_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('scheduled_issue_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    voided_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('voided_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    